lib/linked_list.rb

```diff
@@ -2,44 +2,72 @@
 
 class LinkedList
   attr_reader :head
-
+  
   def initialize
     @head = nil
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if @head.nil?
+    @head.data
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    length = 0
+    curr = @head
+    until curr.nil? do
+      length += 1
+      curr = curr.next
+    end
+    return length
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    if @head.nil?
+      add_first(data)
+    else
+      curr = @head
+      until curr.next.nil? do
+        curr = curr.next
+      end
+      curr.next = Node.new(data, nil)
+    end
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
-
+    curr = @head
```

Comment:
What if `@head` is `nil`?

```suggestion
    return nil if @head.nil?
    curr = @head
```